impeller/toolkit/android/hardware_buffer.cc

```diff
@@ -0,0 +1,105 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/toolkit/android/hardware_buffer.h"
+
+#include "impeller/base/validation.h"
+
+namespace impeller::android {
+
+static AHardwareBuffer_Format ToAHardwareBufferFormat(
+    HardwareBufferFormat format) {
+  switch (format) {
+    case HardwareBufferFormat::kR8G8B8A8UNormInt:
+      return AHARDWAREBUFFER_FORMAT_R8G8B8A8_UNORM;
+  }
+  FML_UNREACHABLE();
+}
+
+static AHardwareBuffer_Desc ToAHardwareBufferDesc(
+    const HardwareBufferDescriptor& desc) {
+  AHardwareBuffer_Desc ahb_desc = {};
+  ahb_desc.width = desc.size.width;
+  ahb_desc.height = desc.size.height;
+  ahb_desc.format = ToAHardwareBufferFormat(desc.format);
+  ahb_desc.layers = 1u;
+  if (desc.usage & static_cast<HardwareBufferUsage>(
+                       HardwareBufferUsageFlags::kFrameBufferAttachment)) {
+    ahb_desc.usage |= AHARDWAREBUFFER_USAGE_GPU_FRAMEBUFFER;
+  }
+  if (desc.usage & static_cast<HardwareBufferUsage>(
+                       HardwareBufferUsageFlags::kCompositorOverlay)) {
+    ahb_desc.usage |= AHARDWAREBUFFER_USAGE_COMPOSER_OVERLAY;
+  }
+  if (desc.usage & static_cast<HardwareBufferUsage>(
+                       HardwareBufferUsageFlags::kSampledImage)) {
+    ahb_desc.usage |= AHARDWAREBUFFER_USAGE_GPU_SAMPLED_IMAGE;
+  }
+  return ahb_desc;
+}
+
+bool HardwareBufferDescriptor::IsAllocatable() const {
+  const auto desc = ToAHardwareBufferDesc(*this);
+  return GetProcTable().AHardwareBuffer_isSupported(&desc) != 0u;
+}
+
+HardwareBuffer::HardwareBuffer(HardwareBufferDescriptor descriptor)
+    : descriptor_(descriptor),
+      android_descriptor_(ToAHardwareBufferDesc(descriptor_)) {
+  if (!descriptor_.IsAllocatable()) {
+    VALIDATION_LOG << "The hardware buffer descriptor is not allocatable.";
+    return;
+  }
+  const auto& proc_table = GetProcTable();
+
+  AHardwareBuffer* buffer = nullptr;
+  if (auto result =
+          proc_table.AHardwareBuffer_allocate(&android_descriptor_, &buffer);
+      result != 0 || buffer == nullptr) {
+    VALIDATION_LOG << "Could not allocate hardware buffer. Error: " << result;
+    return;
+  }
+  buffer_.reset(buffer);
+  is_valid_ = true;
+}
+
+HardwareBuffer::~HardwareBuffer() = default;
+
+bool HardwareBuffer::IsValid() const {
+  return is_valid_;
+}
+
+AHardwareBuffer* HardwareBuffer::GetHandle() const {
+  return buffer_.get();
+}
+
+HardwareBufferDescriptor HardwareBufferDescriptor::MakeForSwapchainImage(
+    const ISize& size) {
+  HardwareBufferDescriptor desc;
+  desc.format = HardwareBufferFormat::kR8G8B8A8UNormInt;
+  // Zero sized hardware buffers cannot be allocated.
+  desc.size = size.Max(ISize{1u, 1u});
```

Comment:
This seems like it should be an error state (std::option or statusOr) instead of returning a size that wasn't requested.

Reply:
I documented that restriction in the headerdoc with a large warning sign so it shows up more prominently. It took me some trial and error to understand that this was the cause of buffers not being allocatable and the Android API doesn't say why a size wasn't allocatable. So making the callers figure this out was a pretty poor experience. Also, we do the same clamping in other spots (like the pbuffer sizes for offscreen GLES surfaces). So it seemed like in keeping with the theme. There are IsAllocatable checks either way though.